misk-crypto/src/main/kotlin/misk/crypto/pgp/RealPgpDecrypter.kt

```diff
@@ -0,0 +1,70 @@
+package misk.crypto.pgp
+
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+import org.bouncycastle.openpgp.PGPCompressedData
+import org.bouncycastle.openpgp.PGPEncryptedDataList
+import org.bouncycastle.openpgp.PGPException
+import org.bouncycastle.openpgp.PGPLiteralData
+import org.bouncycastle.openpgp.PGPOnePassSignatureList
+import org.bouncycastle.openpgp.PGPPrivateKey
+import org.bouncycastle.openpgp.PGPPublicKeyEncryptedData
+import org.bouncycastle.openpgp.PGPUtil
+import org.bouncycastle.openpgp.jcajce.JcaPGPObjectFactory
+import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyDataDecryptorFactoryBuilder
+import org.bouncycastle.util.io.Streams
+import java.security.Security
+
+internal class RealPgpDecrypter(private val privateKey: PGPPrivateKey) : PgpDecrypter {
+  init {
+    Security.addProvider(BouncyCastleProvider())
+  }
+
+  override fun decrypt(byteArray: ByteArray): ByteArray {
+    val buffer = okio.Buffer()
+    val outputStream = buffer.outputStream()
+    val decoderStream = PGPUtil.getDecoderStream(byteArray.inputStream())
+    val jcaPGPObjectFactory = JcaPGPObjectFactory(decoderStream)
+
+    val pgpEncryptedDataList = when (val nextObject = jcaPGPObjectFactory.nextObject()) {
+      is PGPEncryptedDataList -> nextObject
+      else -> jcaPGPObjectFactory.nextObject() as PGPEncryptedDataList
+    }
+
+    val pgpPublicKeyEncryptedData =
+        pgpEncryptedDataList.encryptedDataObjects.next() as PGPPublicKeyEncryptedData
+
+    if (pgpPublicKeyEncryptedData.keyID != privateKey.keyID) {
+      error("this private key won't be able to decrypt this data due to differing key IDs")
+    }
+
+    val publicKeyDataDecryptorFactory = JcePublicKeyDataDecryptorFactoryBuilder()
+        .setProvider("BC")
+        .setContentProvider("BC")
+        .build(privateKey)
+
+    val clear = pgpPublicKeyEncryptedData.getDataStream(publicKeyDataDecryptorFactory)
+    val plainFactory = JcaPGPObjectFactory(clear)
+    var message = plainFactory.nextObject()
+    if (message is PGPCompressedData) {
+      val pgpFactory = JcaPGPObjectFactory(message.dataStream)
+      message = pgpFactory.nextObject()
+    }
+
+    when (message) {
+      is PGPLiteralData -> {
+        Streams.pipeAll(message.inputStream, outputStream)
+        outputStream.close()
+      }
+
+      is PGPOnePassSignatureList ->
+        throw PGPException("encrypted message contains a signed message - not literal data.")
+
+      else -> throw PGPException("message is not a simple encrypted file - type unknown.")
+    }
+    if (pgpPublicKeyEncryptedData.isIntegrityProtected && !pgpPublicKeyEncryptedData.verify()) {
+      error("message failed integrity check")
+    }
+
+    return buffer.readByteArray()
```

Comment:
Please make sure all streams used in this method are properly closed and that buffers that might contain plaintext are overwritten (for example using `bytes.fill(0)`